WordPress/gutenberg
# Editor: Sync blocks state to edited post content

~Blocked by: #9287~ (merged)
~Fixes regression introduced in #9288 (https://github.com/WordPress/gutenberg/pull/9288#issuecomment-416150437)~ (Fixed separately by #9448)
Related: #7970

This pull request seeks to refactor how blocks state is kept in sync with the edited post content, in an effort to simplify / improve consistency of behaviors around editor content initialization.

**Status:** This pull request is rather large and includes a few changes that could potentially be forked out into their own, blocking pull requests (see commits as standalone indicators). It is otherwise expected to be fully functional and passing all tests. New end-to-end tests are planned to verify associated code editor impact, but there are presently issues with the local Docker installation, making end-to-end test authoring impossible for the moment.

**Goal:** The editor module needn't be so conscious of distinct start and end points. It should render itself as derived by its state at any given moment in time. This should help enable efforts like that of #7970, where the editor may need to present itself without relying on a synchronous initialization setup. It is made challenging by various factors such as: enacting a template, assigning initial edits, alerting the user to the presence of an existing autosave which can be restored. To me, these are considerations which should be moved into the scope of `edit-post`. This has its own set of challenges, and is proposed to be addressed separately to the work here.

**Changes:**

- Deprecating `setupEditor`, `setupEditorState`, and `checkTemplateValidity` actions from the editor module
   - Template validation occurs automatically as an effect of `RESET_BLOCKS`
- Refactoring `overridePost` as representing an initial set of edits (for demo content, auto-draft edits)
   - This also avoids needing to account for the auto-draft initial title edit in the editor effects
- Avoid initial history for ignored types in `withHistory` higher-order reducer
   - cc @iseulde , I cannot recall the original reason for this behavior, and while it's sensible to have the initial value to revert back upon as the "starting" history, it doesn't follow that ignored types option would be disregarded to assign this initial value. As it impacts the changes here, without the change it was causing an initial Undo option to be shown when loading the editor.
- Fix bug with how we consider prop and state relationship of edited content value in `PostTextEditor`

**Implementation notes:**

- This changes how we consider the source of truth for post content.
   - Previously there was an awkward juggling of `edits`, where if `content` was assigned as `edits` (as in by the Code Editor textarea), it would be preferred. This was to accommodate the fact that parsing content is a non-performant task to perform on each key press, thus the parse was deferred. However, this introduced a dual source of truth for content (sometimes as a serialization of blocks state, sometimes as the `edits.content` property).
   - With these changes, blocks are always considered as the source of truth for content, and the state value is synced automatically to any changes which occur to the post's edited or persisted content string (via the newly-introduced middleware).
   - To maintain existing Code Editor behavior, an optional flag is supported in the sync middleware to skip the parse. It's assumed this would be followed by a later deferred parse; in the code editor, on the blur of the text field.
      - However, this has a few consequences on how the Code Editor behaves:
         - If the post was not saveable because it had no title, content, or excerpt, and I add content in the Code Editor, it will not be reflected in the UI (by presence of the save button) until I blur the field. 
         - If autosave occurs while I am within Code Editor, the saved payload will not include the in-progress edits, but rather the value of content before I began editing. If I then blur the field, the editor will correctly show as having changes needing to be saved (the deferred parse). One exception is if I press Cmd+R (reload) while within the code editor, I may not be prompted about unsaved changes because the deferred parse has not yet occurred by that time. This could probably be remedied by enhancements proposed by #7409.
      - Considering the goal in consolidating how we consider post content, these changes aren't _technically_ unexpected, but have the potential for a negative impact on user experience.
- This introduces an affordance of "quiet" state changes. We already have a number of state mechanisms and initialization steps which cause state values to change, but for which we don't want to prompt the user about unsaved changes / add history levels. This can be seen with demo content and the auto-draft initial edits, where the initial edits should not be considered as "unsaved", nor should they be undo-able.

**Future tasks:**

- Consider moving more behaviors which continue to live in `EditorProvider` into the `edit-post` module, namely template synchronization and the autosave warning.
   - It's unclear whether "template" should be considered a setting of the editor, as there is mixed expectations on when it is synchronized. Is it just a starting point for a new post? Do we apply the template when editing an existing post which does not adhere to the template? This has the potential for being destructive, and if we were to consider changing it to a user behavior, it would seem to follow that the editor doesn't need to be aware of some ongoing template, since the enacting of the template synchronization would be an explicit action.

**Testing instructions:**

Verify there are no regressions in the initialization of the editor, notably:

- New post without template
   - Should not show "Auto Draft" title
   - Should not save an "Auto Draft" title, even if title is left empty
- New post with template
- Editing existing post with content
- Editing existing post where template would exist, but content differs from template (see also testing instructions from #9288 )
- Demo post

Verify that HTML edits made in the Code Editor are respected:

- Post should be marked as dirty _after focus leaves the textarea field_
- Switching back to Visual Editor should display blocks as edited in Code Editor
